core/src/main/kotlin/org/evomaster/core/problem/external/service/httpws/ExternalServiceHandler.kt

```diff
@@ -109,38 +112,79 @@ class ExternalServiceHandler {
         }
     }
 
+    /**
+     * Reset all the served requests.
+     * The WireMock instances will still be up and running
+     */
+    fun resetServedRequests() {
+        externalServices.forEach { it.value.resetServedRequests() }
+    }
+
     /**
      * This takes all the served requests from WireMock server and creates them
      * as ExternalServiceAction. It ignores if the same absolute URL is added
      * already.
      *
-     * This is not perfect yet, have to explore more scenarios and perfect the
-     * implementation.
+     * This gets reset each time a new main action is evaluated.
      */
-    fun getExternalServiceActions(): MutableList<ExternalServiceAction> {
-        val actions = mutableListOf<ExternalServiceAction>()
+    fun getAllExternalServiceActions(): List<HttpExternalServiceAction> {
+        val actions = mutableListOf<HttpExternalServiceAction>()
         externalServices.forEach { (_, u) ->
             u.getAllServedRequests().forEach {
-                // TODO: This needs to be revised to make it nicer
-                if (externalServiceRequests.none { r -> r.absoluteURL == it.absoluteURL }) {
-                    externalServiceRequests.add(it)
-                }
                 if (actions.none { a -> a.request.url == it.url }) {
-                    actions.add(
-                        ExternalServiceAction(
-                            it,
-                            "",
-                            u,
-                            counter++
-                        )
+                    val action = HttpExternalServiceAction(
+                        it,
+                        "",
+                        u,
+                        counter++
                     )
+                    action.doInitialize(randomness)
+                    actions.add(action)
                 }
-
             }
         }
         return actions
     }
 
+    /**
+     * Return a list of [HttpExternalServiceAction]s for the given absoluteURL.
+     *
+     * There is a possibility to have multiple requests made for the same URL several,
+     * so it returns a list of actions.
+     */
+    fun getExternalServiceActionsForURL(url: String): List<HttpExternalServiceAction> {
```

Comment:
this method is used to create external service actions for a particular rest call.
there could exist multiple requests made for the same URL, but are they all related to the rest call?

Reply:
I'm not sure. If the core is multithreaded there is a possibility. Otherwise, all will be related to the same REST call.

Reply:
a bit unsure.  if there exist different rest actions which access the same URL, one made two requests to the URL, and the other made one. Here, the input argument is the URL. how do you know how many external actions we should return? In addition, if it is multithreaded, do we create two wiremock instances for the same URL at the same time?

Reply:
We don't create two WireMock, but I was thinking like what if same REST call invoked twice.

Reply:
If there do not exist multiple WireMock instances for the same URL, in case that you create two external service actions, then the first one will never work since the 2nd will override the response. right?

`same REST call invoked twice`, does it indicate the REST action? 
if yes, we will have two REST actions in the individual, each of them is under `EnterpiseGroup`, then we could create the corresponding external actions for each.
If here it indicates the call to external service, based on my understanding, we do not support a sequence of responses for the same URL yet, e.g., give different responses based on the sequence  (could support it in the other branch as we discussed). 
in case there exist two external service actions referring to the same URL, how will you handle it now?

Reply:
I'm not so sure about the scenario as I mentioned in my previous comment. Right now there will only one action for the call. Need to check what will happen if there is a query parameter in the request.